File: app/src/main/java/com/antonio/samir/meteoritelandingsspots/service/local/MeteoriteNasaService.java
package com.antonio.samir.meteoritelandingsspots.service.local;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.Observer;
import android.content.Context;
import android.location.Location;
import android.support.annotation.Nullable;

import com.antonio.samir.meteoritelandingsspots.model.Meteorite;
import com.antonio.samir.meteoritelandingsspots.service.repository.MeteoriteRepositoryFactory;
import com.antonio.samir.meteoritelandingsspots.service.repository.database.MeteoriteDao;
import com.antonio.samir.meteoritelandingsspots.service.server.nasa.MeteoriteNasaAsyncTaskService;
import com.antonio.samir.meteoritelandingsspots.service.server.nasa.NasaService;
import com.antonio.samir.meteoritelandingsspots.service.server.nasa.NasaServiceFactory;
import com.antonio.samir.meteoritelandingsspots.util.GPSTracker;

import java.util.ArrayList;
import java.util.List;
import java.util.SortedSet;
import java.util.TreeSet;

class MeteoriteNasaService implements MeteoriteService {

    private final GPSTracker mGpsTracker;
    private final Context mContext;


    public MeteoriteNasaService(final Context context, final GPSTracker gpsTracker) {
        mContext = context;
        mGpsTracker = gpsTracker;
    }

    @Override
    public LiveData<List<Meteorite>> getMeteorites() {

        //Return meteorites
        final MeteoriteDao meteoriteDao = MeteoriteRepositoryFactory.getMeteoriteDao(mContext);

        final boolean gpsEnabled = mGpsTracker.isGPSEnabled();

        final MutableLiveData<List<Meteorite>> list = new MutableLiveData<>();

        final LiveData<List<Meteorite>> liveData = meteoriteDao.getMeteoriteOrdened();

        liveData.observeForever(new Observer<List<Meteorite>>() {

            @Override
            public void onChanged(@Nullable List<Meteorite> meteorites) {

                if (meteorites.isEmpty()) {
                    //If it is empty so load the data from internet
                    final NasaService nasaService = NasaServiceFactory.getNasaService();
                    new MeteoriteNasaAsyncTaskService(nasaService, meteoriteDao).execute();
                } else {
                    //Is not empty so remove the observer
                    list.setValue(meteorites);
                    liveData.removeObserver(this);
                }

                if (gpsEnabled) {
                    final MutableLiveData<Location> mGpsTrackerLocation = mGpsTracker.getLocation();
                    mGpsTrackerLocation.observeForever(new Observer<Location>() {
                        @Override
                        public void onChanged(@Nullable Location location) {
                            if (location != null) {
                                final double latitude = location.getLatitude();
                                final double longitude = location.getLongitude();
                                //String sortOrder = String.format("ABS(reclat - %s ) + ABS(reclong - %s) ASC", latitude, longitude);

                                final SortedSet<Meteorite> sortedSet = new TreeSet<>((m1, m2) -> {
                                    final Double distance1 = m1.distance(latitude, longitude);
                                    final Double distance2 = m2.distance(latitude, longitude);
                                    return (distance1 > 0 && distance2 > 0) ? distance1.compareTo(distance2) : 0;
                                });

                                sortedSet.addAll(meteorites);

                                final ArrayList<Meteorite> value = new ArrayList<>(sortedSet);
                                list.setValue(value);

                                mGpsTrackerLocation.removeObserver(this);
                                mGpsTracker.stopUpdates();

                            }
                        }
                    });

                }

            }
        });

        return list;

    }

}

File: app/src/main/java/com/antonio/samir/meteoritelandingsspots/service/local/MeteoriteNasaService.kt
package com.antonio.samir.meteoritelandingsspots.service.local

import android.location.Location
import androidx.lifecycle.LiveData
import androidx.lifecycle.MediatorLiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Observer
import com.antonio.samir.meteoritelandingsspots.model.Meteorite
import com.antonio.samir.meteoritelandingsspots.service.repository.MeteoriteRepositoryInterface
import com.antonio.samir.meteoritelandingsspots.util.GPSTrackerInterface
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import java.util.*
import java.util.concurrent.atomic.AtomicBoolean

class MeteoriteNasaService(
        private val meteoriteRepository: MeteoriteRepositoryInterface,
        private val addressService: AddressServiceInterface,
        private val gpsTracker: GPSTrackerInterface
) : MeteoriteService {

    private val isUpdateRequired = AtomicBoolean(false)


    override fun loadMeteorites(): LiveData<List<Meteorite>> {

        val result = MediatorLiveData<List<Meteorite>>()

        val liveData = meteoriteRepository.meteoriteOrdened()

        result.addSource(liveData) { meteorites ->

            if (meteorites.isEmpty() && !isUpdateRequired.getAndSet(true)) {
                //If it is empty so load the data from internet
                recoverFromNetwork()
            } else {
                if (gpsTracker.isGPSEnabled()) {
                    gpsTracker.startLocationService()
                    orderByLocation(meteorites, result)
                } else {
                    result.value = meteorites
                }
                addressService.recoveryAddress()
            }

        }

        return result

    }

    override fun addressStatus(): MutableLiveData<String> {
        return addressService.status
    }

    private fun orderByLocation(meteorites: List<Meteorite>, list: MediatorLiveData<List<Meteorite>>) {
        val trackerLocation = gpsTracker.location
        trackerLocation.observeForever(object : Observer<Location> {
            override fun onChanged(location: Location?) {
                if (location != null) {
                    val latitude = location.latitude
                    val longitude = location.longitude
                    //String sortOrder = String.format("ABS(reclat - %s ) + ABS(reclong - %s) ASC", latitude, longitude);

                    val sortedSet = TreeSet<Meteorite> { m1, m2 ->
                        val distance1 = m1.distance(latitude, longitude)
                        val distance2 = m2.distance(latitude, longitude)
                        if (distance1 > 0 && distance2 > 0) distance1.compareTo(distance2) else 0
                    }

                    sortedSet.addAll(meteorites)

                    val value = ArrayList(sortedSet)
                    list.value = value

                    trackerLocation.removeObserver(this)
                    gpsTracker.stopUpdates()

                }
            }
        })
    }

    private fun recoverFromNetwork() {
        GlobalScope.launch(Dispatchers.Default) {
            val remoteMeteorites = meteoriteRepository.getRemoteMeteorites()
            remoteMeteorites?.let { meteoriteRepository.insertAll(it) }
            addressService.recoveryAddress()
        }
    }

    override fun getMeteoriteById(id: String): LiveData<Meteorite>? {
        return meteoriteRepository.getMeteoriteById(id)
    }

}
